arch/cortex-m4/src/mpu.rs

```diff
@@ -116,160 +128,493 @@ impl MPU {
+                region_start += region_size - (region_start % region_size);
+            }
+
+            num_subregions_used = {
+                if initial_kernel_memory_size == 0 {
+                    8
+                } else {
+                    initial_app_memory_size * 8 / region_size + 1
+                }
+            };
+        }
+
+        // Make sure the region fits in the unallocated memory.
+        if region_start + region_size
+            > (unallocated_memory_start as usize) + unallocated_memory_size
+        {
+            return None;
+        }
+
+        // For example: 11111111 & 11111110 = 11111110 --> Use only the first subregion (0 = enable)
+        let subregion_mask = (0..num_subregions_used).fold(!0, |res, i| res & !(1 << i)) & 0xff;
+
+        let region = CortexMRegion::new(
+            region_start as *const u8,
+            region_size,
+            region_start as *const u8,
+            region_size,
+            APP_MEMORY_REGION_NUM,
+            Some(subregion_mask),
+            permissions,
+        );
+
+        config.regions[APP_MEMORY_REGION_NUM] = region;
+
+        Some((region_start as *const u8, region_size))
     }
 
-    fn set_mpu(&self, region: Region) {
-        let regs = &*self.0;
+    fn update_app_memory_region(
+        &self,
+        app_memory_break: *const u8,
+        kernel_memory_break: *const u8,
+        permissions: mpu::Permissions,
+        config: &mut Self::MpuConfig,
+    ) -> Result<(), ()> {
+        let (region_start, region_size) = match config.regions[APP_MEMORY_REGION_NUM].location() {
+            Some((start, size)) => (start as usize, size),
+            None => panic!(
+                "Error: Process tried to update app memory MPU region before it was created."
+            ),
+        };
+
+        let app_memory_break = app_memory_break as usize;
+        let kernel_memory_break = kernel_memory_break as usize;
+
+        // Out of memory
+        if app_memory_break > kernel_memory_break {
+            return Err(());
+        }
+
+        let app_memory_size = app_memory_break - region_start;
+        let kernel_memory_size = region_start + region_size - kernel_memory_break;
+
+        // Determine the number of subregions to enable.
+        let num_subregions_used = {
+            if kernel_memory_size == 0 {
+                8
+            } else {
+                app_memory_size * 8 / region_size + 1
+            }
+        };
+
+        let subregion_size = region_size / 8;
+        let subregions_end = region_start + subregion_size * num_subregions_used;
+
+        // If we can no longer cover app memory with an MPU region without overlapping kernel
+        // memory, we fail.
+        if subregions_end > kernel_memory_break {
+            return Err(());
+        }
 
-        regs.region_base_address.set(region.base_address());
+        let subregion_mask = (0..num_subregions_used).fold(!0, |res, i| res & !(1 << i)) & 0xff;
```

Comment:
Really, create a function that calculates subregion mask. Third occurrence of this!